eval/topdown_test.go

```diff
@@ -0,0 +1,540 @@
+            {"x": "e", "y": 1, "i": 0},
+            {"x": "e", "y": 1, "i": 1},
+            {"x": "e", "y": 1, "i": 2},
+            {"x": "e", "y": 1, "i": 3}
+        ]`},
+		{"p[x] :- d[x][y] = z[i]", `[]`},
+	}
+
+	data := loadSmallTestData()
+
+	for i, tc := range tests {
+
+		ctx := &TopDownContext{
+			Rule:     parseRule(tc.rule),
+			Store:    NewStorageFromJSONObject(data),
+			Bindings: make(map[opalog.Var]opalog.Value),
+		}
+
+		expected := loadExpectedBindings(tc.expected)
+
+		err := evalTerms(ctx, func(ctx *TopDownContext) error {
+			if len(expected) > 0 {
+				for j, exp := range expected {
+					if reflect.DeepEqual(exp, ctx.Bindings) {
+						tmp := expected[:j]
+						expected = append(tmp, expected[j+1:]...)
+						return nil
+					}
+				}
+			}
+			// If there was not a matching expected binding, treat this case as a failure.
+			return fmt.Errorf("unexpected bindings: %v", ctx.Bindings)
+		})
+		if err != nil {
+			t.Errorf("Test case %d: expected success but got error: %v", i+1, err)
+			continue
+		}
+		if len(expected) > 0 {
+			t.Errorf("Test case %d: missing expected bindings: %v", i+1, expected)
+		}
+	}
+}
+
+func TestPlugValue(t *testing.T) {
+
+	a := opalog.Var("a")
+	b := opalog.Var("b")
+	c := opalog.Var("c")
+	k := opalog.Var("k")
+	v := opalog.Var("v")
+	cs := parseTerm("[c]").Value
+	ks := parseTerm(`{k: "world"}`).Value
+	vs := parseTerm(`{"hello": v}`).Value
+	hello := opalog.String("hello")
+	world := opalog.String("world")
+
+	ctx1 := &TopDownContext{Bindings: make(Bindings)}
+	ctx1 = ctx1.Bind(a, b)
+	ctx1 = ctx1.Bind(b, cs)
+	ctx1 = ctx1.Bind(c, ks)
+	ctx1 = ctx1.Bind(k, hello)
+
+	ctx2 := &TopDownContext{Bindings: make(Bindings)}
+	ctx2 = ctx2.Bind(a, b)
+	ctx2 = ctx2.Bind(b, cs)
+	ctx2 = ctx2.Bind(c, vs)
+	ctx2 = ctx2.Bind(v, world)
+
+	expected := parseTerm(`[{"hello": "world"}]`).Value
+
+	r1 := plugValue(a, ctx1.Bindings)
+
+	if !expected.Equal(r1) {
+		t.Errorf("Expected %v but got %v", expected, r1)
+		return
+	}
+
+	r2 := plugValue(a, ctx2.Bindings)
+
+	if !expected.Equal(r2) {
+		t.Errorf("Expected %v but got %v", expected, r2)
+	}
+}
+
+func TestTopDownScalarDoc(t *testing.T) {
+	tests := []struct {
+		note     string
+		rule     string
+		expected string
+	}{
+		{"undefined", "p = null :- false", ""}, // "" will be converted to Undefined
+		{"null", "p = null :- true", "null"},
+		{"bool: true", "p = true :- true", "true"},
+		{"bool: false", "p = false :- true", "false"},
+		{"number: 3", "p = 3 :- true", "3"},
+		{"number: 3.0", "p = 3.0 :- true", "3.0"},
+		{"number: 66.66667", "p = 66.66667 :- true", "66.66667"},
+		{`string: "hello"`, `p = "hello" :- true`, `"hello"`},
+		{`string: ""`, `p = "" :- true`, `""`},
+		{"array: [1,2,3,4]", "p = [1,2,3,4] :- true", "[1,2,3,4]"},
+		{"array: []", "p = [] :- true", "[]"},
+		{`object/nested composites: {"a": [1], "b": [2], "c": [3]}`,
+			`p = {"a": [1], "b": [2], "c": [3]} :- true`,
+			`{"a": [1], "b": [2], "c": [3]}`},
```

Comment:
This test is poorly named, it should be "TestTopDownComplete" or something. We don't support set literals yet.